### PR TITLE
Added a pile of docs to expression.py

### PR DESCRIPTION
**Pull request by tpoterba**
Sorry @catoverdrive, the RNG really likes you recently.

Despite being 1700 lines of changes, I don't think this is a huge commitment to review.
(only a couple of non-doc changes, which I'll point out)

### CODE REVIEW

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 1]
```diff
@@ -971,6 +971,9 @@ object FunctionRegistry {
         result: "7"
     """,
     "s" -> "String of the form ``CHR:POS:REF:ALT`` or ``CHR:POS:REF:ALT1,ALT2...ALTN`` specifying the contig, position, reference and alternate alleles.")(stringHr, variantHr(GR))
+
+  register("AltAllele", { (ref: String, alt: String) => AltAllele(ref, alt)}, "")(stringHr, stringHr, altAlleleHr)
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: This was totally missing

File: python/hail/genetics/variant.py [hunk 2]
```diff
@@ -307,7 +307,7 @@ def __str__(self):
         return self._jrep.toString()
 
     def __repr__(self):
-        return 'AltAllele(ref=%s, alt=%s)' % (self.ref, self.alt)
+        return "AltAllele(ref='{ref}', alt='{alt}')".format(ref=self.ref, alt=self.alt)
```
Comment (tpoterba) [inline python/hail/genetics/variant.py]: Much more pythonic

Comment (tpoterba) [reply python/hail/genetics/variant.py]: (having repr return strings wrapped in '')

File: python/hail/expr/expression.py [hunk 3]
```diff
@@ -87,6 +87,10 @@ def to_expr(e):
         return construct_expr(ApplyMethod('Locus', Literal('"{}"'.format(str(e)))), TLocus(e.reference_genome))
     elif isinstance(e, Interval):
         return construct_expr(ApplyMethod('Interval', Literal('"{}"'.format(str(e)))), TInterval(e.reference_genome))
+    elif isinstance(e, Call):
```
Comment (tpoterba) [inline python/hail/expr/expression.py]: This logic was totally missing, so Call/AltAllele couldn't be interpreted as strings

File: python/hail/expr/expression.py [hunk 4]
```diff
@@ -1010,11 +1745,33 @@ def _set_field(self, key, value):
...
+        -------
+        :class:`Expression`
+            Struct field.
+        """
+
+        if item in self._fields:
+            return self._fields[item]
         else:
-            raise AttributeError("Could not find field `" + str(item) + "' in schema.")
+            raise AttributeError("No field '{}' in schema. Available fields: {}".format(
```
Comment (tpoterba) [inline python/hail/expr/expression.py]: this was a bug (the self.__dict__ thing)

File: python/hail/expr/expression.py [hunk 5]
```diff
@@ -1240,19 +2346,63 @@ class Int64Expression(NumericExpression):
...
+
+        Returns
+        -------
+        :class:`StringExpression`
+            Substring or character at index `item`.
+        """
         if isinstance(item, slice):
             return self._slice(TString(), item.start, item.stop, item.step)
-        elif isinstance(item, int):
+        elif isinstance(item, int) or isinstance(item, Int32Expression):
```
Comment (tpoterba) [inline python/hail/expr/expression.py]: bugfix here too

File: python/hail/expr/expression.py [hunk 6]
```diff
@@ -1240,19 +2346,63 @@ class Int64Expression(NumericExpression):
...
+        ----------
+        other : :class:`StringExpression`
+            String to concatenate.
+
+        Returns
+        -------
+        :class:`StringExpression`
+            Concatenated string.
+        """
+        other = to_expr(other)
```
Comment (tpoterba) [inline python/hail/expr/expression.py]: Bugfix (previously if `other` was just a python str, we'd get a NotImplementedError)

**Comment by tpoterba** [comment]
OK, I think those are all the code changes

File: python/hail/expr/expression.py (hunk 5 above)
Comment (catoverdrive) [reply python/hail/expr/expression.py]: what is self._index? can you just pass it a expression like that? (also can a slice have an expression in it?)

Comment (tpoterba) [reply python/hail/expr/expression.py]: Here are the definitions of those two guys:
```python
    def _index(self, ret_typ, key):
        key = to_expr(key)
        indices, aggregations, joins = unify_all(self, key)
        return construct_expr(Index(self._ast, key._ast),
                              ret_typ, indices, aggregations, joins)

    def _slice(self, ret_typ, start=None, stop=None, step=None):
        if start is not None:
            start = to_expr(start)
            start_ast = start._ast
        else:
            start_ast = None
        if stop is not None:
            stop = to_expr(stop)
            stop_ast = stop._ast
        else:
            stop_ast = None
        if step is not None:
            raise NotImplementedError('Variable slice step size is not currently supported')

        non_null = [x for x in [start, stop] if x is not None]
        indices, aggregations, joins = unify_all(self, *non_null)
        return construct_expr(Index(self._ast, Slice(start_ast, stop_ast)),
                              ret_typ, indices, aggregations, joins)
```

Comment (tpoterba) [reply python/hail/expr/expression.py]: You can indeed pass an expression into an index as well as a slice!

File: python/hail/expr/expression.py [hunk 7]
```diff
@@ -1289,20 +2459,34 @@ def replace(self, pattern1, pattern2):
...
             Delimiter regex.
-        n : int or :class:`Int32Expression`, optional
+        n : :class:`Int32Expression`, optional
             Maximum number of splits.
 
         Returns
         -------
-        :class:`ArrayStringExpression`
+        :class:`ArrayExpression`
+            Array of split strings.
```
Comment (konradjk) [inline python/hail/expr/expression.py]: Malformed regex wiki link under `matches()` below

File: python/hail/expr/expression.py [hunk 8]
```diff
@@ -1601,243 +2942,454 @@ def category(self):
...
 
     @property
     def alt_alleles(self):
         """Returns the alternate alleles in the polymorphism.
 
+        Examples
+        --------
+        .. doctest::
+
+            >>> eval_expr(variant.alt_allele())
```
Comment (konradjk) [inline python/hail/expr/expression.py]: `alt_alleles`

File: python/hail/expr/expression.py [hunk 9]
```diff
@@ -341,13 +413,25 @@ def exists(self, f):
         Returns
         -------
         :py:class:`BooleanExpression`.
-            `True` if `f` returns `True` for any element, `False` otherwise.
+            ``True`` if `f` returns ``True`` for any element, ``False`` otherwise.
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/expression.py]: found some more too!

File: python/hail/expr/expression.py [hunk 10]
```diff
@@ -518,6 +710,13 @@ def median(self):
     def product(self):
         """Returns the product of all elements in the collection.
 
+        Examples
+        --------
+        .. doctest::
+
+            >>> eval_expr(a.product())
+            120
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/expression.py]: yes totally

File: python/hail/expr/expression.py [hunk 11]
```diff
@@ -679,41 +994,164 @@ def _bin_op_numeric_reverse(self, name, other):
...
-    def __div__(self, other):
-        return self._bin_op("/", other, TArray(TFloat64()))
+        Examples
+        --------
+        .. doctest::
 
-    def __rdiv__(self, other):
-        return self._bin_op_reverse("/", other, TArray(TFloat64()))
+            >>> eval_expr(a2 * 5)
+            >>> eval_expr(a2 * 5)
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/expression.py]: copy/pasted the wrong thing!

File: python/hail/expr/expression.py (hunk 8 above)
Comment (tpoterba) [reply python/hail/expr/expression.py]: eyes of a hawk

**Pull request merged**